src/lib/mdns/minimal/Server.cpp

```diff
@@ -0,0 +1,242 @@
+                                  chip::Inet::InterfaceId interface)
+{
+    for (size_t i = 0; i < mEndpointCount; i++)
+    {
+        EndpointInfo * info = &mEndpoints[i];
+        if (info->udp == nullptr)
+        {
+            continue;
+        }
+
+        if (info->addressType != addr.Type())
+        {
+            continue;
+        }
+
+        chip::Inet::InterfaceId boundIf = info->udp->GetBoundInterface();
+
+        if ((boundIf != INET_NULL_INTERFACEID) && (boundIf != interface))
+        {
+            continue;
+        }
+
+        return info->udp->SendTo(addr, port, data);
+    }
+
+    chip::System::PacketBuffer::Free(data);
+    return CHIP_ERROR_NOT_CONNECTED;
+}
+
+CHIP_ERROR ServerBase::BroadcastSend(chip::System::PacketBuffer * data, uint16_t port)
+{
+    for (size_t i = 0; i < mEndpointCount; i++)
+    {
+        EndpointInfo * info = &mEndpoints[i];
+
+        if (info->udp == nullptr)
+        {
+            continue;
+        }
+
+        // data may be sent over multiple packets. Keep the one ref active all the time
+        data->AddRef();
+
+        CHIP_ERROR err;
+
+        if (info->addressType == chip::Inet::kIPAddressType_IPv6)
+        {
+            err = info->udp->SendTo(kBroadcastIp.ipv6, port, info->udp->GetBoundInterface(), data);
+        }
+        else if (info->addressType == chip::Inet::kIPAddressType_IPv4)
+        {
+            err = info->udp->SendTo(kBroadcastIp.ipv4, port, info->udp->GetBoundInterface(), data);
+        }
+        else
+        {
+            // remove extra ref and then also clear it
+            chip::System::PacketBuffer::Free(data);
+            chip::System::PacketBuffer::Free(data);
+            return CHIP_ERROR_INCORRECT_STATE;
+        }
+
+        if (err == chip::System::MapErrorPOSIX(ENETUNREACH))
+        {
+            // Send attempted to an unreachable network. Generally should not happen if
+            // interfaces are configured properly, however such a failure to broadcast
+            // may not be critical either.
+            ChipLogError(Discovery, "Attempt to mDNS broadcast to an unreachable destination.");
+        }
+        else if (err != CHIP_NO_ERROR)
+        {
+            chip::System::PacketBuffer::Free(data);
+            return err;
+        }
+    }
+
+    chip::System::PacketBuffer::Free(data);
+    return CHIP_NO_ERROR;
+}
+
+void ServerBase::OnUdpPacketReceived(chip::Inet::IPEndPointBasis * endPoint, chip::System::PacketBuffer * buffer,
+                                     const chip::Inet::IPPacketInfo * info)
+{
+    chip::System::AutoFreePacketBuffer autoFree(buffer);
+
+    ServerBase * srv = static_cast<ServerBase *>(endPoint->AppState);
+    if (!srv->mDelegate)
+    {
+        return;
+    }
+    mdns::Minimal::BytesRange data(mdns::Minimal::BytesRange(buffer->Start(), buffer->Start() + buffer->DataLength()));
```

Comment:
remove this redundant copy?

Reply:
Heh ... copy and paste got to me. Fixed.